File: python/questions/averageOfLevels.py
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        lns = []
        ans = []
        lns.append(root)
        cls = len(lns)
        while cls:
            cs = 0
            for p in range(cls):
                cn = lns[0]
                lns.pop(0)
                if cn.left:
                    lns.append(cn.left)
                if cn.right:
                    lns.append(cn.right)
                cs += cn.val
            ans.append(cs / cls)
            cls = len(lns);
        return ans

File: cpp/questions/averageOfLevels.cpp
#include <vector>
#include <queue>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

using namespace std;

class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        std::queue<TreeNode*> ln;                       
        std::vector<double> ans;
        ln.push(root);
        while (!ln.empty()) {
            int cls = ln.size();
            double cs = 0;
            for (int p=0; p<cls; p++) {
                TreeNode* cn = ln.front();
                if (cn->left) {
                    ln.push(cn->left);
                }
                if (cn->right) {
                    ln.push(cn->right);
                }
                ln.pop();
                cs += cn->val;
            }
            ans.push_back(cs / cls);
        }
        return ans;
    }
};
